### PR TITLE
Fix polarity of NodePort logic to avoid leaked ports

### LINKED ISSUE #43233: Nodeport assigned to two services on 3 Master HA Cluster
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.):

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Kubernetes version** (use `kubectl version`): K8S 1.5.3, and kubectl 1.5.3

**Environment**:
- **Cloud provider or hardware configuration**: azure
- **OS** (e.g. from /etc/os-release): xenial
- **Kernel** (e.g. `uname -a`): latest xenial
- **Install tools**:
- **Others**: ACS-Engine cluster, 3 masters

**What happened**:
Nodeport 30177 is assigned to dashboard and to the nginx service that was started within a minute of starting the dashboard.  The dashboard is being started as part of the addon manager.

**What you expected to happen**:
Each service should receive a separate node port.

**How to reproduce it** (as minimally and precisely as possible):
Run 3 master setup, with 3 API servers, 3 addon managers, 3 controllers (but one controller is elected leader).  Run dashboard as an add-on.  Create service shortly after dashboard is started an assigned nodeport.

We run a Jenkin's job, and this repros about 30-50% of the time. 

**Anything else we need to know**:
We instrumented the build and found the following:
1. Dashboard is initially created by the addon manager by issuing a POST to/api/v1/namespaces/kube-system/services.  This is where port 30177 is added, and snapshotted to etcd
2.  Immediately after the POST, another addon manager issues PATCH /api/v1/namespaces/kube-system/services/kubernetes-dashboard.  The post calls a release to 30177, and this is snapshotted to etcd.
3. Another API server receives the request for nginx and assigns 30177 to nginx service.
4. Afterwards all api servers start reporting message the node port 30177 for service nginx/namespace-15616 was assigned to multiple services.

**Workaround**: delete the nginx service, re-create the service.

@thockin have you seen this before?  This build has your latest changelist around repair.go #39206, but from best we can observe is unrelated to the changes.

### PR DESCRIPTION
**Pull request by thockin**
The result of this was that an update to a Service would release the
NodePort temporarily (the repair loop would fix it in a minute).  During
that window, another Service could get allocated that Port.

Fixes #43233

### CODE REVIEW

**Comment by thockin** [comment]
@k8s-bot verify test this

**Comment by thockin** [comment]
@k8s-bot gci gce e2e test this

**Comment by thockin** [comment]
@k8s-bot gci gce e2e test this

**Comment by thockin** [comment]
User confirms this fixed the problem.  Once LGTM I will offer a 1.5.x patch

**Comment by brendandburns** [comment]
/lgtm

Thanks for the quick fix @thockin

**Comment by smarterclayton** [comment]
Any way to simulate this race from a unit test?

**Comment by thockin** [comment]
yeah, I need to get a test ready for this, but I didn't want to block getting this up.

File: pkg/registry/core/service/rest.go [hunk 1]
```diff
@@ -436,7 +436,7 @@ func (rs *REST) Update(ctx genericapirequest.Context, name string, objInfo rest.
 	// The comparison loops are O(N^2), but we don't expect N to be huge
 	// (there's a hard-limit at 2^16, because they're ports; and even 4 ports would be a lot)
 	for _, oldNodePort := range oldNodePorts {
-		if !contains(newNodePorts, oldNodePort) {
+		if contains(newNodePorts, oldNodePort) {
```
Comment (justinsb) [inline pkg/registry/core/service/rest.go]: Gah - I should stop doing double-negatives :-(

**Comment by colemickens** [comment]
@thockin do we not need to cherry-pick this into release-1.6 as well?

**Comment by thockin** [comment]
shoot, so many balls in the air - YES, we need t pick this into 1.5 and
1.6, and I don't even know the current process for this stage of the
release cycle.

@ethernetdan ?

**Comment by colemickens** [comment]
Let me know... I can go ahead and generate and send the cherry-pick for 1.6... it looks like [you got it into 1.5 already](https://github.com/kubernetes/kubernetes/pull/43270).

**Comment by thockin** [comment]
Is it the same process for 1.6 release branch?

**Comment by thockin** [comment]
+++ Creating a pull request on GitHub at
thockin:automated-cherry-pick-of-#43259-upstream-release-1.6
Error creating pull request: Unprocessable Entity (HTTP 422)
No commits between kubernetes:release-1.6 and
thockin:automated-cherry-pick-of-#43259-upstream-release-1.6


Did it get bulk merged?

**Comment by colemickens** [comment]
I don't really understand how, but your commit is in release-1.6...

I manually checked the commit history and looked at the current state of the code...

Sorry for the noise, I guess it's all already taken care of.

**Comment by ethernetdan** [comment]
@thockin @colemickens we are still fastforwarding changes into release-1.6 from master. Will stop after rc.1 comes out or the release, whichever comes first.

**Comment by thockin** [comment]
OK.  but merge queue is only taking milestoned PRs, right? :)

**Comment by ethernetdan** [comment]
It is but it's implemented as a [blacklist](https://github.com/enisoc/test-infra/blob/e287770/mungegithub/submit-queue/deployment/kubernetes/configmap.yaml#L115). We probably want to change that 😃

**Comment by thockin** [comment]
yeah, that should be fixed.

**Pull request merged**